native/dsterf.go

```diff
@@ -0,0 +1,283 @@
+	}
+	if len(e) < n-1 {
+		panic(badE)
+	}
+
+	// Determine the unit roundoff for this environment.
+	eps := dlamchE
+	eps2 := eps * eps
+	safmin := dlamchS
+	safmax := 1 / safmin
+	ssfmax := math.Sqrt(safmax) / 3
+	ssfmin := math.Sqrt(safmin) / eps2
+
+	// Compute the eigenvalues of the tridiagonal matrix.
+	maxit := 30
+	nmaxit := n * maxit
+	jtot := 0
+
+	l1 := 0
+
+	var anorm, c, gamma, r, rte, s, sigma float64
+	var iscale, l, lend, lendsv, lsv, m int
+
+Ten:
+	if l1 > n-1 {
+		goto OneSeventy
+	}
+	if l1 > 0 {
+		e[l1-1] = 0
+	}
+	for m = l1; m < n-1; m++ {
+		if math.Abs(e[m]) <= math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1]))*eps {
+			e[m] = 0
+			goto Thirty
+		}
+	}
+	m = n - 1
+
+Thirty:
+	l = l1
+	lsv = l
+	lend = m
+	lendsv = lend
+	l1 = m + 1
+	if lend == 0 {
+		goto Ten
+	}
+
+	// Scale submatrix in rows and columns l to lend.
+	anorm = impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+	iscale = 0
+	if anorm == 0 {
+		goto Ten
+	}
+	if anorm > ssfmax {
+		iscale = 1
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l, 1, e[l:], n)
+	} else if anorm < ssfmin {
+		iscale = 2
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l, 1, e[l:], n)
+	}
+
+	for i := l; i < lend; i++ {
+		e[i] *= e[i]
+	}
+
+	// Choose between QL and QR iteration.
+	if math.Abs(d[lend]) < math.Abs(d[l]) {
+		lend = lsv
+		l = lendsv
+	}
+	if lend >= l {
+		// QL Iteration.
+		// Look for small subdiagonal element.
+	Fifty:
+		if l != lend {
+			for m = l; m < lend; m++ {
+				if math.Abs(e[m]) <= eps2*(math.Abs(d[m]*d[m+1])) {
+					goto Seventy
+				}
+			}
+		}
+		m = lend
+	Seventy:
+		if m < lend {
+			e[m] = 0
+		}
+		p := d[l]
+		if m == l {
+			goto Ninety
+		}
+		// If remaining matrix is 2 by 2, use Dlae2 to compute its eigenvalues.
+		if m == l+1 {
+			rte = math.Sqrt(e[l])
+			rt1, rt2 := impl.Dlae2(d[l], rte, d[l+1])
+			d[l] = rt1
+			d[l+1] = rt2
+			e[l] = 0
+			l += 2
+			if l <= lend {
+				goto Fifty
+			}
+			goto OneFifty
+		}
+		if jtot == nmaxit {
+			goto OneFifty
+		}
+		jtot++
+
+		// Form shift
+		rte = math.Sqrt(e[l])
+		sigma = (d[l+1] - p) / (2 * rte)
+		r = impl.Dlapy2(sigma, 1)
+		sigma = p - (rte / (sigma + math.Copysign(r, sigma)))
+
+		c = 1.0
+		s = 0
+		gamma = d[m] - sigma
+		p = gamma * gamma
+
+		// Inner loop.
+		for i := m - 1; i >= l; i-- {
+			bb := e[i]
+			r := p + bb
+			if i != m-1 {
+				e[i+1] = s * r
+			}
+			oldc := c
+			c = p / r
+			s = bb / r
+			oldgam := gamma
+			alpha := d[i]
+			gamma = c*(alpha-sigma) - s*oldgam
+			d[i+1] = oldgam + (alpha - gamma)
+			if c != 0 {
+				p = (gamma * gamma) / c
+			} else {
+				p = oldc * bb
+			}
+		}
+		e[l] = s * p
+		d[l] = sigma + gamma
+		goto Fifty
+
+	Ninety:
+		// Eigenvalue found.
+		d[l] = p
+		l++
+		if l <= lend {
+			goto Fifty
+		}
+		goto OneFifty
+	} else {
+	OneHundred:
```

Comment:
I think so. Added TODO. Once dsyev is in I'll have more confidence in correctness and will feel more at ease with modifying complex goto statements.